src/TodoContext/TodoContext.tsx

```diff
@@ -0,0 +1,121 @@
+import React, { useEffect, useMemo, useState } from 'react';
+import { filterForTodos } from '../services/chooseFilterForTodos';
+import { getTodos } from '../api/todos';
+import { USER_ID } from '../variables/UserID';
+import { Todo } from '../types/Todo';
+import { TodoContext } from '../types/TodoContext.1';
+import { wait } from '../utils/fetchClient';
+
+export const TodosContext = React.createContext<TodoContext>({
+  todos: [],
+  addTodo: () => { },
+  setCompleted: () => { },
+  makeAllCompleted: () => { },
+  query: '',
+  setQuery: () => { },
+  filteredTodos: [],
+  deleteCompletedTodos: () => { },
+  deleteTodo: () => { },
+  saveEditingTitle: () => { },
+  errorMessage: '',
+  setErrorMessage: () => { },
+});
+
+interface Props {
+  children: React.ReactNode;
+}
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [query, setQuery] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Error');
+
+        wait(3000).then(() => setErrorMessage(''));
+      });
+  }, []);
+
+  function addTodo(todo: Todo) {
+    if (todo.title.trim().length) {
+      setTodos(prevTodos => [
+        ...prevTodos,
+        todo,
+      ]);
+    }
+  }
+
+  function setCompleted(todoID: number) {
+    const changeCompletedTodos = todos.map(todo => {
+      return todo.id === todoID
+        ? { ...todo, completed: !todo.completed }
+        : todo;
+    });
+
+    setTodos(changeCompletedTodos);
+  }
+
+  const filteredTodos = filterForTodos(query, todos);
+
+  function saveEditingTitle(todoID: number, changedTitle: string) {
+    let changedTodos = [...todos];
+
+    if (changedTitle.trim()) {
+      changedTodos = changedTodos.map(todo => (
+        todo.id === todoID
+          ? { ...todo, title: changedTitle }
+          : todo));
+    }
+
+    setTodos(changedTodos);
+  }
```

Comment:
bad way to do it with `let`
Analyze it & make improvements in all cases
```suggestion
    const isValid = changedTitle.trim();
    
    const getNewTodos = (prevTodos: Todo[]) => {
      if (isValid) {
        return prevTodos.map(todo => (
          todo.id === todoID
            ? { ...todo, title: changedTitle }
            : todo));
      } else {
        return prevTodos;
      }
      
    setTodos(saveTodos);
  }
```

OR

```suggestion
    const isValid = changedTitle.trim();
    
    setTodos(prevTodos: Todo[]) => {
      if (isValid) {
        return prevTodos.map(todo => (
          todo.id === todoID
            ? { ...todo, title: changedTitle }
            : todo));
      } else {
        return prevTodos;
      }
  }
```